### PR TITLE
[WEB-671]: Setup TriggerCapiEventInput GraphQL Mutation

### PR DESCRIPTION
**Pull request by scottkicks**
# 📲 What

Before we can send CAPI events, we need to implement a mutation that will allow us to send the relevant data to our server.
This mutation will be called in the following cases when:
1. Members have given consent for KS to track their data
2. The relevant project has it's `sendMetaCapiEvents` property set to true

**Event Cases:**
- Project page visit
- Select back this project
- Initiate checkout (selects a reward)
- Add payment info tapped (could be swapping existing card or adding anew card)
- Purchase (have reached the thank you page)

# 🤔 Why

If Facebook app users decline permission for FB to track them for advertising purposes, then our KS Facebook Pixel will be rendered useless. Using Facebook's Conversions API (CAPI) is the recommended workaround. 

# 🛠 How

The web team will be handling the actual events to Facebook's Conversions API and has added this new mutation so that we can notify them about what data and which events to send. 

# ✅ Acceptance criteria

- [x] Network call is readily available and usable.

### CODE REVIEW

**Comment by nativeksr** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: KsApi/Service.swift [hunk 1]
```diff
@@ -106,6 +106,21 @@ public struct Service: ServiceType {
       .flatMap(CreatePaymentSourceEnvelope.producer(from:))
   }
 
+  public func triggerCapiEventInput(input: TriggerCapiEventInput)
+    -> SignalProducer<EmptyResponseEnvelope, ErrorEnvelope> {
+    return GraphQL.shared.client
+      .perform(mutation: GraphAPI
+        .TriggerCapiEventMutation(input: GraphAPI
+          .TriggerCapiEventInput(
```
Comment (msadoon) [inline KsApi/Service.swift:114]: inject input directly here, following examples above.

File: KsApi/mutations/inputs/ApplePayParams.swift [hunk 2]
```diff
@@ -1,20 +0,0 @@
-import Foundation
-
-public struct ApplePayParams: Encodable, Equatable {
```
Comment (msadoon) [inline KsApi/mutations/inputs/ApplePayParams.swift:3]: eh why is this and the tests deleted?

File: KsApi/MockService.swift [hunk 3]
```diff
@@ -1561,6 +1566,23 @@
...
 
+    internal func triggerCapiEvent(input: TriggerCapiEventInput)
+      -> SignalProducer<EmptyResponseEnvelope, ErrorEnvelope> {
+      guard let client = self.apolloClient else {
+        return .empty
+      }
+
+      let mutation = GraphAPI
+        .TriggerCapiEventMutation(input: GraphAPI
+          .TriggerCapiEventInput(
```
Comment (msadoon) [inline KsApi/MockService.swift:1577]: again, pass in as just an input, no need to pass in every property.

File: KsApi/Service.swift [hunk 4]
```diff
@@ -106,6 +106,21 @@ public struct Service: ServiceType {
...
+            projectId: input.projectId,
+            eventName: input.eventName,
+            externalId: input.externalId
+          )))
+      .flatMap { _ in
```
Comment (msadoon) [inline KsApi/Service.swift:119]: No need create the closure with `_ in`, just using `producer(from:))` like above.

File: KsApi/ServiceType.swift [hunk 5]
```diff
@@ -47,6 +47,10 @@ public protocol ServiceType {
   func addImage(file fileURL: URL, toDraft draft: UpdateDraft)
     -> SignalProducer<UpdateDraft.Image, ErrorEnvelope>
 
+  /// Views project after navigating to the via facebook ad deep-link
```
Comment (msadoon) [inline KsApi/ServiceType.swift:50]: Watch comments here, ensure they are correctly written for each function and reflect its use.

File: Kickstarter.xcodeproj/project.pbxproj [hunk 6]
```diff
@@ -483,6 +483,12 @@
...
 		6067BCF2293FC3520036ABB1 /* FacebookResetPasswordViewModelTests.swift in Sources */ = {isa = PBXBuildFile; fileRef = 6067BCEF293FC10E0036ABB1 /* FacebookResetPasswordViewModelTests.swift */; };
 		606F214429799A1200BA5CDF /* ATTrackingAuthorizationStatus.swift in Sources */ = {isa = PBXBuildFile; fileRef = 606F214229799A0000BA5CDF /* ATTrackingAuthorizationStatus.swift */; };
+		606F214C2994158500BA5CDF /* TriggerCapiEventInput.swift in Sources */ = {isa = PBXBuildFile; fileRef = 606F214B2994158500BA5CDF /* TriggerCapiEventInput.swift */; };
+		606F214E2994237400BA5CDF /* AppDataInput.swift in Sources */ = {isa = PBXBuildFile; fileRef = 606F214D2994237400BA5CDF /* AppDataInput.swift */; };
+		606F2151299423E100BA5CDF /* CustomDataInput.swift in Sources */ = {isa = PBXBuildFile; fileRef = 606F2150299423E100BA5CDF /* CustomDataInput.swift */; };
+		606F215429943EB300BA5CDF /* AppDataInputTests.swift in Sources */ = {isa = PBXBuildFile; fileRef = 606F215229943DAE00BA5CDF /* AppDataInputTests.swift */; };
+		606F215729943FA200BA5CDF /* CustomDataInputTests.swift in Sources */ = {isa = PBXBuildFile; fileRef = 606F215529943F9400BA5CDF /* CustomDataInputTests.swift */; };
```
Comment (msadoon) [inline Kickstarter.xcodeproj/project.pbxproj:490]: Why are these files added to the `.pbx` file but not shown in this pr?

**Review by msadoon — CHANGES_REQUESTED** [review]
Left some review comments, let's go over in our 1-1, feel free to correct before our meeting too :)

File: KsApi/Service.swift (hunk 4 above)
Comment (scottkicks) [reply KsApi/Service.swift:114]: Looking at the other examples that use `EmptyResponseEnvelope`, they do create this closure with a SignalProducer. I believe this is because `EmptyResponseEnvelope` doesn't already have a static `producer(from:)` function

**Review by msadoon — CHANGES_REQUESTED** [review]
Almost perfect, just add a test class for `TriggerCapiEventInput.swift` and use the example of `CreatePaymentSourceSetupIntentInputTests` to write a quick test.

**Comment by scottkicks** [comment]
@msadoon not sure why some files are being added/not added by git. The original test file that I had already done should be in here now. Sorry about that.

File: KsApi/mutations/inputs/TriggerCapiEventInputTests.swift [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+      )
+
+    let input = triggerCapiEventInput.toInputDictionary()
+
+    XCTAssertEqual(input["projectId"] as? String, "projId")
+    XCTAssertEqual(input["eventName"] as? String, "eventName")
+    XCTAssertEqual(input["externalId"] as? String, "extId")
+    XCTAssertEqual(input["userEmail"] as? String, "userEmail")
+    XCTAssertNotNil(input["appData"])
+    XCTAssertNotNil(input["customData"])
```
Comment (scottkicks) [inline KsApi/mutations/inputs/TriggerCapiEventInputTests.swift:23]: Since `GraphAPI.AppDataInput` & `GraphAPI.CustomDataInput` don't conform to Equatable _and_ they're auto generated for us, I opted to just assert that the are not nil here.

**Review by msadoon — APPROVED** [review]
This looks fine for now. I didn't test it via Charles, but it's probably better if you get Charles successfully logging calls so you can see the results for yourself or in the future with new network calls.

There's a bunch of tutorials online. One I found here seems pretty detailed - https://www.kodeco.com/21931256-charles-proxy-tutorial-for-ios

If you want we can pair to configure this on your mac.
